### PR TITLE
fix: ensure search params are not left off of `load_cache`

### LINKED ISSUE #10122: beforeNavigate hook loses search parameters present in the URL after successive navigations
### Describe the bug

I am trying to use `beforeNavigate` to preserve a query parameter in the URL if it is present. The important bit is as follows, in the root `+layout.svelte` file:

```svelte
<script>
  import { beforeNavigate } from "$app/navigation";
  const paramName = "test";
  beforeNavigate(({ from, to }) => {
    const previousParamValue = from?.url.searchParams.get(paramName);
    if (typeof previousParamValue === "string") {
      to?.url.searchParams.set(paramName, previousParamValue);
    }
  })
</script>

<slot />
```

Now, if I run `npm run dev` and visit `localhost:5173?test=true` and start navigating between pages, the `test` param will be preserved for the first navigation, but not for navigations after that. This is because the `from` value passed to the `beforeNavigate` callback does not contain the query parameters that are in the actual URL for navigations after the first one following the query parameter being added.

The `navigating` store from `$app/stores` exhibits the same problematic behavior.

The workaround is relatively simple:

```svelte
<script>
  import { beforeNavigate } from "$app/navigation";
  import { page } from "$app/stores";
  const paramName = "test";
  $: lastPageURL = $page?.url;
  beforeNavigate(({ to }) => {
    const previousParamValue = lastPageURL?.searchParams.get(paramName);
    if (typeof previousParamValue === "string") {
      to?.url.searchParams.set(paramName, previousParamValue);
    }
  })
</script>

<slot />
```

### Reproduction

Repro repository: https://github.com/Benaiah/repro-svelte-navigation-hook-bug

Instructions:

- Clone the repo
- `npm install`
- `npm run dev`
- Visit the dev site url (e.g. `localhost:5173`)
- Click the button at the bottom of the page to add a query parameter to the URL
- Navigate around the app and watch the query parameter disappear on the second navigation

The logs attached are a copy of the console logs in the browser as this process is followed.

### Logs

```Shell
$page.url: http://localhost:5173/page-2 +layout.svelte:7:13
afterNavigate: undefined → http://localhost:5173/page-2 +layout.svelte:24:12
beforeNavigate: http://localhost:5173/page-2 → http://localhost:5173/page-2?preserveThisParam=true +layout.svelte:18:12
$navigating: http://localhost:5173/page-2 → http://localhost:5173/page-2?preserveThisParam=true +layout.svelte:9:12
$page.url: http://localhost:5173/page-2?preserveThisParam=true +layout.svelte:7:13
afterNavigate: http://localhost:5173/page-2 → http://localhost:5173/page-2?preserveThisParam=true +layout.svelte:24:12
beforeNavigate: http://localhost:5173/page-2?preserveThisParam=true → http://localhost:5173/?preserveThisParam=true +layout.svelte:18:12
$navigating: http://localhost:5173/page-2?preserveThisParam=true → http://localhost:5173/?preserveThisParam=true +layout.svelte:9:12
$page.url: http://localhost:5173/?preserveThisParam=true +layout.svelte:7:13
afterNavigate: http://localhost:5173/page-2?preserveThisParam=true → http://localhost:5173/?preserveThisParam=true +layout.svelte:24:12
beforeNavigate: http://localhost:5173/ → http://localhost:5173/page-1 +layout.svelte:18:12
$navigating: http://localhost:5173/ → http://localhost:5173/page-1 +layout.svelte:9:12
$page.url: http://localhost:5173/page-1 +layout.svelte:7:13
afterNavigate: http://localhost:5173/ → http://localhost:5173/page-1 +layout.svelte:24:12
```

### System Info

```Shell
System:
    OS: macOS 13.4
    CPU: (8) arm64 Apple M1 Pro
    Memory: 88.88 MB / 32.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.12.1 - ~/.nvm/versions/node/v18.12.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v18.12.1/bin/yarn
    npm: 8.19.2 - ~/.nvm/versions/node/v18.12.1/bin/npm
  Browsers:
    Chrome: 114.0.5735.106
    Firefox: 113.0.2
    Safari: 16.5
  npmPackages:
    @sveltejs/adapter-auto: ^2.0.0 => 2.1.0 
    @sveltejs/kit: ^1.5.0 => 1.20.2 
    svelte: ^3.54.0 => 3.59.1 
    vite: ^4.3.0 => 4.3.9
```

### Severity

serious, but I can work around it

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by iteratetograceness**
Closes #10122. 

The `load_cache` has no context of any search parameters that exist on the current path. So when `data-sveltekit-preload-data` is enabled for links, search parameters were being left off in the `navigate` function. More specifically, when setting the `current` variable which is used to inform the `from` property in `BeforeNavigate`. 

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [x] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Review by saeeddirect1 — APPROVED** [review]

File: packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<script>
+	import { beforeNavigate } from '$app/navigation';
+
+    // do not change, used in test (/basics/test/cross-platform/client.test.js):
+    const preserved_param = 'preserved';
+    const preserved_param_value = 'true';
```
Comment (ghostdevv) [inline packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte:6]:
```suggestion
```

This doesn't seem to be used, unless it was intended for line 11?

File: packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<script>
```
Comment (ghostdevv) [inline packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte:1]: The formatting seems a bit messed up in this file, can you run prettier in it?

File: packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte (hunk 1 above)
Comment (jueungrace) [reply packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte:6]: @ghostdevv yes, you're totally right!

File: packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte (hunk 2 above)
Comment (jueungrace) [reply packages/kit/test/apps/basics/src/routes/before-navigate/preserve-search-params/+layout.svelte:1]: @ghostdevv good catch, actually not sure why it hadn't caught the double-quotes the first time! :')

**Review by ghostdevv — DISMISSED** [review]
LGTM, hopefully another maintainer can take a look soon to give feedback :pray:

**Comment by jueungrace** [comment]
I will go bug @tcc-sejohnson for a review!

**Comment by tcc-sejohnson** [comment]
Hmmm, I'm not sure if this is quite right... if I'm understanding correctly, preloading `/foo?bar=baz` and then clicking a link for `/foo` would still navigate to `/foo?bar=baz`. Maybe I'm wrong -- I'm reading code on mobile. I'll dig in further when I get home.

**Comment by jueungrace** [comment]
@tcc-sejohnson this was actually a bit confusing at first for me too! 

I'll try my best at a better description:

We're not preloading links that have a search param but _without_ (`/foo`); when we navigate to `/foo`, what's saved in the `load_cache` is `/foo`. When we have logic inside a `beforeNavigate` to persist a search param using `to` and `from`, the search param ends up getting lost after the first navigation because the value for `to` gets set via the `current` variable -- and `current` this gets set using the `load_cache` (which in this example currently holds `/foo` not something like `/foo?bar=baz`). 

Please do let me know if you see a better way to tackle this! For a better understanding of the bug, it'll help to start from the repo in the issue linked or the added test.

**Comment by dummdidumm** [comment]
Thanks for tackling this!

I think the fix is applied at a point that is too late. If the the search param is appended, a `load` function dealing with it should see that search param, too - which is not the case with this fix. So I think instead the `load_cache` should be ignored completely if a mismatch is detected. The root problem is that the `id` which determines this is computed in `get_navigation_intent` which is before calling `beforeNavigate` which could change the search params and therefore the `id`.

In general I'm wondering if we actually _want_ to fix this. what would happen if someone changes the pathname in `beforeNavigate`? Then we're at a completely different page. Should we handle that? Should we instead document that these parameters should be treated as readonly and tell people to instead do something like this?

```js
beforeNavigate(({ from, to, cancel }) => {
	const previous_param_value = from?.url.searchParams.get(preserved_param);
	if (typeof previous_param_value === 'string') {
		const url = new URL(to.url);
		url.searchParams.set(preserved_param, 'asd');
		cancel();
		goto(url);
	}
});
```

Then again, `beforeNavigate` has no info about any of the additional `goto` parameters (like `noScroll`), and it would also need additional checks to handle `goto/invalidate` etc correctly. So I could also be persuaded to say "search params are a special case which we handle". I'm not sure what the correct approach here is yet.

**Comment by jueungrace** [comment]
@dummdidumm really appreciate you taking a look!

Based on what you've shared, I'm also wondering whether this is something we want to explicitly fix -- and whether there's an appropriate "fix" for this based on how/the order of navigation and preloading are currently handled. 

I can take some more time this week/next to dig a little deeper on the flow of navigation and see if there may be a better approach!

**Comment by Rich-Harris** [comment]
preview: https://svelte-dev-git-preview-kit-10179-svelte.vercel.app/

_this is an automated message_

**Comment by eltigerchino** [comment]
Closing this since I can't reproduce this original issue after upgrading to the latest version of SvelteKit.

**Pull request closed without merging**